### ISSUE TITLE
Problem with matching by set distance

### ISSUE THREAD

**Issue by KalelR** [JuliaDynamics/Attractors.jl#143]
I'd like some help in figuring out what's going wrong in a synthetic example I created. I'm trying to apply global_continuation using AttractorsViaFeaturizing, first with a matching by set distance. The system's real behavior is as described in the left column of the figure below.

![matching_by_flow_example-method-2-dummy_multistable_equilibrium!](https://github.com/user-attachments/assets/09895543-9fcf-4d29-b74a-82a1d6532189)

The first attractor, which starts at position 0, moves to -5 for params 3 and 3.5, but moves back to 0 at param>=4.  Matching by centroid misses this behavior, and 'incorrectly' matches the attractors, as shown in the right column.

I had run these results before the reformulation of `global_continuation` and everything. Now I am getting an error:
![image](https://github.com/user-attachments/assets/58c286f5-0b1b-47ad-864a-e69d9249490c)

There are missing values of the basin fractions. As far as I understand, the problem is in the `rmap` that you get. This simple example shows what's happening:
```julia 
fs = Dict(4 => 0.4097706879361914, 2 => 0.19042871385842472, 3 => 0.18843469591226322, 1 => 0.21136590229312063)
rmap = Dict(2 => 3) #error
# rmap = Dict(2 => 3, 3=>2) #fine
 
swap_dict_keys!(fs, rmap)
```

The rmap for those parameters is missing one entry (e.g. 3=>2). This makes the swap function drop one of the keys. I am not sure why this is happening, but I know that removing line 112 of `sssdistance.jl`, which is 
```julia 
    filter!(p -> p.first ≠ p.second, rmap)
```

solves it, because then rmap has 2=>3 and 3=>2. I don't know why, though. 
Could you please confirm this is not just on my machine? I've updated everything now.

If you don't want to run the continuation, it all boils down to these guys: 
```julia
pcurve = [[1 => 1.0], [1 => 1.5], [1 => 2.0], [1 => 2.5], [1 => 3.0], [1 => 3.5], [1 => 4.0], [1 => 4.5], [1 => 5.0]]
attractors_cont_simple = Dict{Int64, SVector{1, Float64}}[Dict(1 => [0.0]), Dict(1 => [0.0]), Dict(2 => [2.0], 1 => [0.0]), Dict(2 => [0.0], 1 => [2.0]), Dict(2 => [2.0], 3 => [4.0], 1 => [-5.0]), Dict(2 => [4.0], 3 => [-5.0], 1 => [2.0]), Dict(4 => [6.0], 2 => [0.0], 3 => [2.0], 1 => [4.0]), Dict(4 => [4.0], 2 => [0.0], 3 => [2.0], 1 => [6.0]), Dict(5 => [8.0], 4 => [6.0], 2 => [0.0], 3 => [2.0], 1 => [4.0])]
attractors_cont = [Dict(k=>StateSpaceSet(Vector(v)) for (k,v) in atts) for atts in attractors_cont_simple]
fractions_cont = [Dict(1 => 1.0), Dict(1 => 1.0), Dict(2 => 0.8091908091908092, 1 => 0.19080919080919082), Dict(2 => 0.1906187624750499, 1 => 0.8093812375249501), Dict(2 => 0.18862275449101795, 3 => 0.6207584830339321, 1 => 0.1906187624750499), Dict(2 => 0.6211365902293121, 3 => 0.19042871385842472, 1 => 0.18843469591226322), Dict(4 => 0.4097706879361914, 2 => 0.19042871385842472, 3 => 0.18843469591226322, 1 => 0.21136590229312063), Dict(4 => 0.21115537848605578, 2 => 0.1902390438247012, 3 => 0.18824701195219123, 1 => 0.4103585657370518), Dict(5 => 0.21613545816733068, 4 => 0.1942231075697211, 2 => 0.1902390438247012, 3 => 0.18824701195219123, 1 => 0.21115537848605578)]

rmaps = match_sequentially!(
    attractors_cont, MatchBySSSetDistance(); pcurve, ds = Attractors.referenced_dynamical_system(mapper)
)
match_sequentially!(fractions_cont, rmaps)
```

Note that e.g. `fractions_cont[7]` is now incomplete. The problem seems to be with retract_keys somehow.

Thanks!

**Comment by Datseris** [comment]
yeah, this line:

```
    filter!(p -> p.first ≠ p.second, rmap)
```

is new. I added it to remove entries of the `rmap` that leave things unchanged: if an ID is mapped in the same ID it does not need to exist in the replacement map (makes sense). 

Perhaps I have forgotten to test the case you describe, where the `rmap` should have been `3 ->2, 2=>3` but I am almost certain I tested it. I will see the code. If i understand correctly, I only need the `attractors_cont` you define and then the `match_sequentially` function. I don't need the fractions at all.

**Comment by KalelR** [comment]
rmaps is the same for me. The problem to me happens when you apply it to fractions_cont:

`match_sequentially!(fractions_cont, rmaps)`

**Comment by Datseris** [comment]
I don't get any error and everything works fine for me. Additionally, if you look at `rmaps` above, there is no situation where you have `2 => 3` but not `3 => 2`, which is what you mentioned as the basis of the error.

So, to be clear, the following **Minimal Working Example** yields no errors:
```julia
using Attractors

pcurve = [[1 => 1.0], [1 => 1.5], [1 => 2.0], [1 => 2.5], [1 => 3.0], [1 => 3.5], [1 => 4.0], [1 => 4.5], [1 => 5.0]]
attractors_cont_simple = Dict{Int64, SVector{1, Float64}}[Dict(1 => [0.0]), Dict(1 => [0.0]), Dict(2 => [2.0], 1 => [0.0]), Dict(2 => [0.0], 1 => [2.0]), Dict(2 => [2.0], 3 => [4.0], 1 => [-5.0]), Dict(2 => [4.0], 3 => [-5.0], 1 => [2.0]), Dict(4 => [6.0], 2 => [0.0], 3 => [2.0], 1 => [4.0]), Dict(4 => [4.0], 2 => [0.0], 3 => [2.0], 1 => [6.0]), Dict(5 => [8.0], 4 => [6.0], 2 => [0.0], 3 => [2.0], 1 => [4.0])]
attractors_cont = [Dict(k=>StateSpaceSet(Vector(v)) for (k,v) in atts) for atts in attractors_cont_simple]


mapped_atts = deepcopy(attractors_cont)
matcher = MatchBySSSetDistance()
rmaps = match_sequentially!(mapped_atts, matcher)

fractions_cont = [Dict(1 => 1.0), Dict(1 => 1.0), Dict(2 => 0.8091908091908092, 1 => 0.19080919080919082), Dict(2 => 0.1906187624750499, 1 => 0.8093812375249501), Dict(2 => 0.18862275449101795, 3 => 0.6207584830339321, 1 => 0.1906187624750499), Dict(2 => 0.6211365902293121, 3 => 0.19042871385842472, 1 => 0.18843469591226322), Dict(4 => 0.4097706879361914, 2 => 0.19042871385842472, 3 => 0.18843469591226322, 1 => 0.21136590229312063), Dict(4 => 0.21115537848605578, 2 => 0.1902390438247012, 3 => 0.18824701195219123, 1 => 0.4103585657370518), Dict(5 => 0.21613545816733068, 4 => 0.1942231075697211, 2 => 0.1902390438247012, 3 => 0.18824701195219123, 1 => 0.21115537848605578)]
match_sequentially!(fractions_cont, rmaps)
```

**Comment by KalelR** [comment]
What `fractions_cont` do you get? Here I am getting
```julia
julia> fractions_cont
9-element Vector{Dict{Int64, Float64}}:
 Dict(1 => 1.0)
 Dict(1 => 1.0)
 Dict(2 => 0.8091908091908092, 1 => 0.19080919080919082)
 Dict(2 => 0.8093812375249501, 1 => 0.1906187624750499)
 Dict(2 => 0.18862275449101795, 3 => 0.1906187624750499, 1 => 0.6207584830339321)
 Dict(2 => 0.18843469591226322, 3 => 0.19042871385842472, 1 => 0.6211365902293121)
 Dict(4 => 0.4097706879361914, 3 => 0.19042871385842472, 1 => 0.21136590229312063)
 Dict(4 => 0.4103585657370518, 3 => 0.1902390438247012, 1 => 0.21115537848605578)
 Dict(5 => 0.21613545816733068, 4 => 0.1942231075697211, 3 => 0.1902390438247012, 1 => 0.21115537848605578)
```

note that the 7th entry is missing `2`.

**Comment by Datseris** [comment]
ok, yes, there is bug:
```julia
using Attractors

attractors_cont_simple = Dict{Int64, SVector{1, Float64}}[Dict(1 => [0.0]), Dict(1 => [0.0]), Dict(2 => [2.0], 1 => [0.0]), Dict(2 => [0.0], 1 => [2.0]), Dict(2 => [2.0], 3 => [4.0], 1 => [-5.0]), Dict(2 => [4.0], 3 => [-5.0], 1 => [2.0]), Dict(4 => [6.0], 2 => [0.0], 3 => [2.0], 1 => [4.0]), Dict(4 => [4.0], 2 => [0.0], 3 => [2.0], 1 => [6.0]), Dict(5 => [8.0], 4 => [6.0], 2 => [0.0], 3 => [2.0], 1 => [4.0])]
attractors_cont = [Dict(k=>StateSpaceSet(Vector(v)) for (k,v) in atts) for atts in attractors_cont_simple]


mapped_atts = deepcopy(attractors_cont)
matcher = MatchBySSSetDistance()
rmaps = match_sequentially!(mapped_atts, matcher)

fractions_cont = [Dict(1 => 1.0), Dict(1 => 1.0), Dict(2 => 0.8091908091908092, 1 => 0.19080919080919082), Dict(2 => 0.1906187624750499, 1 => 0.8093812375249501), Dict(2 => 0.18862275449101795, 3 => 0.6207584830339321, 1 => 0.1906187624750499), Dict(2 => 0.6211365902293121, 3 => 0.19042871385842472, 1 => 0.18843469591226322), Dict(4 => 0.4097706879361914, 2 => 0.19042871385842472, 3 => 0.18843469591226322, 1 => 0.21136590229312063), Dict(4 => 0.21115537848605578, 2 => 0.1902390438247012, 3 => 0.18824701195219123, 1 => 0.4103585657370518), Dict(5 => 0.21613545816733068, 4 => 0.1942231075697211, 2 => 0.1902390438247012, 3 => 0.18824701195219123, 1 => 0.21115537848605578)]

mapped_fracs = deepcopy(fractions_cont)
match_sequentially!(mapped_fracs, rmaps)

display(mapped_atts[7])
display(mapped_fracs[7])
display(fractions_cont[7])
display(rmaps[6])
```

indeed the mapped fractions are missing an entry. I don't know the problem cause.

**Comment by Datseris** [comment]
indeed you are right, commenting out this line 
```julia
filter!(p -> p.first ≠ p.second, rmap)
```
fixes the problem.

**Comment by KalelR** [comment]
> indeed the mapped fractions are missing an entry. I don't know the problem cause.

It's something to do when the keys are retracted. It starts this problem with keys being overwritten, but I'm not sure how/why exactly

**Comment by Datseris** [comment]
BUT WHY?!??!?!?!

**Comment by Datseris** [comment]
The code for sure doesn't work as intended because even with the line `filter!(p -> p.first ≠ p.second, rmap)` some rmaps are `Dict(2 => 2)` which is exactly the kind of replacements we want to remove. 

You mentioned retract keys? Why do you think that is a culptrit?

**Comment by KalelR** [comment]
> The code for sure doesn't work as intended because even with the line `filter!(p -> p.first ≠ p.second, rmap)` some rmaps are `Dict(2 => 2)` which is exactly the kind of replacements we want to remove.
> 
> You mentioned retract keys? Why do you think that is a culptrit?

The filter does work, but the retract keys, which comes after it, can reintroduce 2=>2. 

If you make `rmaps = match_sequentially!(mapped_atts, matcher; retract_keys=false)` this problem of overwriting keys does not happen.

**Comment by Datseris** [comment]
But how? 
```julia
function retract_keys_to_consecutive(v::Vector{<:Dict})
    ukeys = unique_keys(v)
    ukeys = setdiff(ukeys, [-1]) # skip key -1 if it exists
    rmap = Dict(k => i for (i, k) in enumerate(ukeys) if i != k)
    return rmap
end
```

it literally says `if i != k`. 

??????????? How is this happening

**Comment by Datseris** [comment]
You have found the bug; if you find how to fix it can you please make a PR for it? I have to do some other stuff now!

**Comment by KalelR** [comment]
I dont know either haha. I'm too tired to continue today, I can take a look again tomorrow and open a PR when I figure it out!

**Comment by Datseris** [comment]
lol

oh my god

it's because in the function `match_sequentially!` the retraction is applied to the `attractors` vector of dictionaries directly.

It is not modifying the `rmaps` instead.

To fix it, we need to make `match_sequentially` use retract keys to modify `rmaps` instead and at its end call `match_sequentially!(attractors ,rmaps)`. 

So, in essence, the `rmaps` we currently return is WRONG if retract keys is true, as retarct keys doesn't apply to rmap.

Please fix it in a pr please please

**Comment by Datseris** [comment]
the solution is after `rmaps` is estimated to apply the key retraction logic to it directly. the `attractors` dictionary in `match_sequentially!` should not be called/used at all inside the `if retract_keys` clause.

So the `rmaps` is modified by retraction and at the end of teh function we simplu call `match_sequentially!(attractors, rmaps)`.

**Comment by KalelR** [comment]
I don't think that solves it, actually. The retraction logic is correct, as far as I see. Even with the attractors. The problem is that applying the retraction onto rmap can sometimes lead to keys missing. 

For instance, rmap might say 4=>7, 2=>5 before retraction. The retraction says 7=>4, 5=>3 and 3=>2.  So after retraction, rmap becomes 4=>4, 2=>3. But, currently, it does not add 3=>2. This leads to 3 being dropped.
I see two options: 
(i) either implement a logic that adds 3=>2 to match_sequentially! I've been trying for some minutes to figure this out but it's not trivial in my head or 
(ii) remove the filter line; without the filter, the 2=>3 is already added.

**Comment by Datseris** [comment]
can you do a pr with this, and use the code you used here as a test suite?

**Comment by Datseris** [comment]
Actually, the filter line can be added at the very end of `match_sequentially!`. Go over all dictionaries in `rmaps` and filter htem with this line. But AFTER the retraction has happened.

**Comment by KalelR** [comment]
yes, but why add this complexity?

**Comment by KalelR** [comment]
This was solved now in PR #144 by removing the filter line.
